text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,801 @@
+
+Using the dictionary above, we can compress to:
+
+```
+std::iter::Chain<$1::Zip<$0::vec::IntoIter<u32>, $6>>
+```
+
+A couple of things to note:
+
+ - The first occurrence of a dictionary entry is never substituted. We don't store the dictionary anywhere and need to be able to reconstruct it from the compressed version.
+ - Longer substitutions are preferred to shorter ones. `std::iter::Chain<$1::Zip<$0::vec::IntoIter<u32>, $4::IntoIter<u32>>>` would also decompress to the original version but the compiler is supposed to always pick the longest substitution available.
+
+The mangled version of a substitution marker is `S <key - 1> _` (and `S_` for key `0`) like in the Itanium mangling. So the above definition would be mangled to:
+
+```
+_RN12std_a1b2c3d44iter5ChainINS0_3ZipINS_3vec8IntoIterIjEES5_EEE
+```
+
+The uncompressed version would be:
+```
+_RN12std_a1b2c3d44iter5ChainIN12std_a1b2c3d44iter3ZipIN12std_a1b2c3d43vec8IntoIterIjEEN12std_a1b2c3d43vec8IntoIterIjEEEEE
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The reference-level explanation consists of three parts:
+
+1. A specification of the syntax mangled names conform to.
+2. A specification of the compression scheme.
+3. A mapping of Rust entities to the mangling syntax.
+
+For implementing a demangler, only the first two sections are of interest, that is, a demangler only needs to understand syntax and compression of names, but it does not have to care about how the compiler generates mangled names.
+
+
+## Syntax Of Mangled Names
+
+The syntax of mangled names is given in extended Backus-Naur form:
+
+ - Non-terminals are within angle brackets (as in `<name-prefix>`)
+ - Terminals are within quotes (as in `"_R"`),
+ - Optional parts are in brackets (as in `[<decimal>]`),
+ - Repetition (zero or more times) is signified by curly braces (as in `{ <name-prefix> }`)
+ - Comments are marked with `//`.
+
+Mangled names conform to the following grammar:
+
+```
+// The <decimal-number> specifies the encoding version.
```

Comment:
Do we need this?

I guess it would be hard to remain backwards-compatible with demanglers when we *add* something, given how arbitrary all the rules are, but that doesn't mean we need to specify a way to upgrade the version, unless we want to change the meaning of the existing syntax?

Do you envision some special handling when a demangler sees a digit after `_R`?

Reply:
A demangler could quickly determine that it cannot handle a given symbol. And it could quickly switch between different demangling algorithms based on the version.

It is a more complicated problem than I thought at first, though. I think we'll always try to change as little as possible when adding something. It still means that symbol using new features can't be demangled by older demanglers.

It's still good to have some version marker in there. If there was a radical change in the encoding, we could still keep the `_R` prefix this way.